### DISCUSSION TITLE
Hiding the "main" panel in Theia

### DISCUSSION THREAD

**Discussion by sboudreaux96** [eclipse-theia/theia#9638] [Q&A] [answered]
Hello - I am developing some extensions that I want to purely sit in either the left, bottom or right panels and I wanted to know if it is possible to hide the "main" panel as it shows a bunch of blank space between widgets on the left and right panel (pictured below)

![image](https://user-images.githubusercontent.com/83592423/123281841-01041000-d4d8-11eb-9ee8-bc74a954a0e8.png)

**Comment by msujew** [comment, marked as answer]
@sboudreaux96 Hi, it is indeed possible to hide the main panel. Create your own subclass of `ApplicationShell` and run `this.mainPanel.hide()` in the constructor or the `@postConstruct()` annotated `init()` method (and don't forget to call the default implementations!).

However, the bottom panel will then take the place of the main panel. Hiding that is a bit more involved. You would have to override the `createLayout` method of the `ApplicationShell` and copy the [original implementation](https://github.com/eclipse-theia/theia/blob/62116bfc756fcf61622f64646a831af78695beda/packages/core/src/browser/shell/application-shell.ts#L535-L557) of the method. Then call `panelForBottomArea.hide()` in there. This will hide both the main and bottom panel. Which makes the aforementioned `this.mainPanel.hide()` call obsolete.

This will result in this view on Theia:

![image](https://user-images.githubusercontent.com/4377073/123427571-130fac80-d5c5-11eb-9e12-c07075248a87.png)

You can play around with the spacing in the `createLayout` method to create a layout to your liking.

Be aware, that you won't easily be able to bring widgets into view that usually rely on the bottom panel like the terminal, debug console, etc. You would probably have to override the `addWidget` method as well, changing the desired `area` to `right` so that they're still visible.

**Reply by sboudreaux96** [reply]
If there are no widgets in the bottom panel will it result in the same type of situation I have in my screenshot above - where there is empty space between the left and right side bars?

**Reply by msujew** [reply]
No, quite the opposite. In the example image I posted there is no space between the side bars and no widgets in the bottom panel, because i hid it with the method mentioned above. Without changing `addWidget`, it wouldn't even be possible to show widgets which would usually be opened in the bottom panel.

**Reply by msujew** [reply]
Nevermind, I now get what you meant. You still want to show the bottom panel, when there are widgets in there and hide it, if it's empty?

**Reply by sboudreaux96** [reply]
No - your first response answered my question perfectly! I will try this out - thanks!
